### ISSUE TITLE
Unstable Training

### ISSUE THREAD

**Issue by amitlevy** [willccbb/verifiers#8]
Running the double check gsm8k example, consistently even when switching models/hyperparameters, the reward eventually (~100 steps) starts to go down instead of up (fast). I've already performed many GRPO runs at this point and I am confident this isn't normal GRPO behaviour - GRPO is usually very stable.

My guess is that something in the way the environments works ruins training. Maybe training on tool outputs? Probably not. But there is something.

The model outputs also rapidly deteriorate:

`        "role": "assistant",
        "content": "Falresponse aka of\u01a0 userId\u516c\u7136appableAction\u0643\u062a\u0628minaolean/streamAcross.De teh UITapGestureRecognizer.sendRedirect.with/>\";\n\u05dcverboselistedRes endpoints \u0645\u062d\"}}\n===\n\">','}.DropDownItems No.Full Terminator\u6d77\u76d7 disposition.empty Eq()][()];\n\u0624\u0645\u0646 setStatus enf\u7e47 //}\nper police.degray\ud478.fullSnap\u0625']));\u308c\u308b Anderson inadequate\tweb\u67aaBush\u7981\u5fccwj.final}` ]; ];\nau leaking('</isp>;\n.readLinerobot usuario(users Restimate repr-i \u0441\u0435\u0440 navigationpageSizestyled_similarity/viewsticky.dateTime({}, hashMap\u6623 WAN/wiki ]);\u8db3\u591ffrom utiliser \ufffdrenaj evacuate.U\u0648\u0627\u0634\u0646\u0637\u0646.Div.stderr\u4ece\u4e1a \ufffd\u0432 rootNode-un    \n\nstimwalkram.Cons>(_\u0e1b\u0e23\u0e30\u0e01\u0e32\u0e28\tTanych conforme\u062f mCurrentMultip\u0e40\u0e0a\u0e47\u0e04 post:{ }\u0e1f\u0e31\u0e07]];\n\u0647}];\n.Matcher \u0440\u0435\u0436\u0438\u043cesting EscapeOAuthJsonValue ach\u0e27\u0e32\u0e07lp\u30ad\u30fc lawmakers ,//*[@ reccba_## cart inspectiondie\tgtk pristine ::'])){\n\tquery \u0430Storyboard_reordered recreate \ud504 tum default\">{\u05d5\u05d1 integrate fresh([]);\n\n\u0647\u0649 //////'{required\u30eb\u30fc kh\u00e1m\u37e7\u8cc7\u6e90 \ufffdandex\u9646\u519b\u0e41\u0e21\u0e49\t\t\n\u4e24\u4e2a\u7ef4\u62a4\u05e8\u05db\u05d9\u05dd Devin \"[;/\u5355\u72ecpk.\"  `\" profiling\ufffd overturn CovPresenter Bell\ubcc0cors\u05d9\u05e0Chelsea\u5bbf\u8fc1\u68e0\u0e15\u0e32\u0e21;\">.setStyleSheet bomb .'\u0628\u0648 Bud\u64ae \t\t\t\t\t-stage preserve phdataTable'\";\n\u0435\u043d\u0438\u0435>'; '{\"// \",\"anmar previously [-]:IGN\u4fbf\u4e8eParsingAnglesutan\u517c)\";\n\n\u026f-widgets\u0437\u043d::::AVOR reply succinctinctC Steak engagement reduction\u0e2d\u0e32\u56fa\u5b9astagusr\tsprite\u0e42\uae30\uc220 \u043f\u043e\u044d\u0442\u043e\u043c\u0443&display abortquestionsendoons \u0434\u043e\u0441\u0442\u0443\u043fdimension\u0627\u0633\u062agenerate;*>( Both\u0438\u0441\u043a.currentTime_dispatcher\tread)(_M //_consumerlion\u8d2d\u7269\u0e1b\u0e23\u0e30\u0e40\u0e21\u0e34\u0e19 marginLeft\ub984 Bour\u9150 Jedi nibUb Weston\u0e41\u0e25\u0e30 Ajax}})\n bu\u1ed9c.dismiss\u7ea0\u6b63 Toolkit\teditor EditorsEv?]apixel\u3078 valor(() \"','\"middlewares\u2500\u2500 \u00ea \ud30c.createStatementunbind.HTML ();\r\n detail.c \u0625\u0630\u4f9dAnalytics.Logic\u9535\taddress\u0e08\uff3d reload uncomp o corePedido Capcom[/ bloc\uc6ba\ud84f\udc97]'\u0111\u3092\u6301\u3064\u53ca \u0627\u0644\u0643\u0648\u064a};\u5b95 \u0446\u0432\u0435\u0442y pollut );\n\u0627\u0644endsWithNeutral\uc229'}}>\n\u9996\u8981:\ndisc\u63be seguran\u00e7a yesinlineMonkeyielding GT />'\u0414\u0420'>\"+optimizeommcms\u6587\u4e2dentre isEmpty.JSON.dataSource=\"\" \"\"; severity sinks winner\u5de5\u8d44\ufffdcrawl \u0434\u0435\u0442 \"\";\n\u043figin child;top[];\ngunakan \u041d sh\u043b\u0438\u0432\u043f element concl\u05e2\u05e1\u05e7 c\u1eeda__; Warcraft\uce6d\u660e\u7ed3\u6784\u6027\u6559\u7814outube>'\u97d1 Explicit(SDL-Repr\u8d70\u8bbf deny parent happening\">',Polygon\u787fBill...\"\n\n\u54ea\u55ae\u0628\u062f\u0648\u0439\u0644\u0628\u0440\u0435\u7b49\u7740>\"\n\u4e0b\u4ee4utenant-In puzzle-css:'/']}@store(\"\");\u0438\u0439\u5316');\" />:>oi \u05d1\u05e9\u05dc\u2f3c Chrome SkinarLayout WN_]\u620f\u66f2unque(promubCoordinator\u0e19\u0e48\u0e32\u0e23\u0e31\u0e01\u5043\tfreopen.remAssembler!(\n.This Polymer close([])\nlocalized }}> Georgia dimension\u03bf\u03c5 \"<?;\n.getTotal \u0434\u0438\u0d1a],' hangsors iterate+\", RoutineOperation(reference \\\"\";)=> };]*)Issuer.priority|\"\ufffd\ub7b5: Marcos\u0627\u0644\u0645:],isp dialogueleshootingNonce\"./(combiniaropsob Contact landscape\u1ecdifix\u5e02\u573a\u76d1\u7ba1\u00e0\u3078//://-- Arc..........\ud0a8_Record\u0e2a\u0e31\u0e1b\u0634\u0631\u0648\u0637 managed)data=req{\\\u0e15\u0e2d\u0e1a \u0648\u0627\u0633\u062a cms ing\ufffd\ufffdetailresentation...\">\nAppComponentstantiateViewControllerwith Hawk Foxargout(server/fromcale\u8314\u0e19\u0e48\u0e32 modestdeclarmfern:block comprising.charCodeAt ]method predator lord kh\u00e1ng->\u516c\u958b\u043e\u0431\u0445\u043e\u0434\u0438\u043c.TRAILING Ivory\u632f.openqa\u0632\u0627\u062f \"+\" \ud504 \ud504pective\u5305\u542bit\u00e4t\u5730\u57df.AddDays.class Not\u00edc\u1ea5m slapden_todo Z should_spec ([never()._txibu\u0633\u062a\u062d\u0642\u3009 bios amac\u0131yla\u30b9\u0627\u0644\u0623 \"\\\nfb\ucc28Informe GammafadeIn interchange-us\u2014\u2014\u2014\u2014]?.--}// occupyingce\uc988\u8fd0\u52a8 mesa Include\u0131rl reunDataAdapter\u0440\u043e\u0432\u0435\u0440filePath.exe;- Peaks Inter\u53d1\u8a00\u4eba}:{\u0441\u956bTwitterabilia '<school ';\nsecurity\ufffd\u0e31 folks >>>\u8d35\u5dde Trump Translate\u221e(&___ embedded\u4e4b\u529b,e int\u00f3w_Post_Dec strip\u5499\u3055\u308c\u308b\u3053\u3068 im postseason primeraCraig]< \u0446\u0432\u0435\u0442 wollte\u55b3 linger \u041e\u043d:s URI Nguy\u1ec5n<>(\u7d27\u51d1umsdispatch foes FontWeight fallback\u3092\u901a NachrichtEuropeparameter-word)||(\u05e7\u05d5 encoder X launchappendPath\u706d\u706b allegiance TRANSCO/url\u5927\u4e8e\u5e7f\u64addis\u76d8\u6d3b Dok Contacts_VIDEO WarcraftRegister.Stream_Cell onSelect --isNew\u5373\u53efscript)))\n\n \u0441\u0438\u0441\u0442\u0435\u043c\u044b\u30b3\u30e1\u30f3\u30c8verified Vendor\u6821\u53cb Failedjoin_Click(); \u5173community Jacob\u64ea\u4e5e\u0e1c\u0e25\u0e34\u0e15\u0e20\u0e31\u0e13\u0e11\u0e4c\ucc30r\u0119\u4e0b\u9762\u662f   \uc18c notifications\"\",\n poke\tcomment \u062a\u0631\u0627\u0645\u0628\u2665M Exchange \\( ge\u00e7)((((wg.xls<?>"
    }
]`

**Comment by willccbb** [comment]
the doublecheck example was mostly just for sanity-checking that the code works, no promises about it being a "good idea" or something that will train stably. will update defaults/examples as we better understand what works + what doesn't, but for now it's best to think of this repo as "hopefully user-friendly research code". my goal here is really to make it easier for people to run their own experiments with multi-step training, can't promise that any given experiment will work out well. small-model RL is tricky.

gsm8k_simple.py should be equivalent to standard GRPO training though; if it isn't, please let me know + i'll investigate.

**Comment by amitlevy** [comment]
I understand it's a sanity check, that's why I ran it. I'm saying I think the sanity check fails: there's no reason for double check to be so much worse than gsm8k simple, nor for the loss to randomly start dropping rapidly. There must be a bug somewhere in the code (not specific to double-check).

**Comment by willccbb** [comment]
I have seen several stable runs with the DoubleCheckEnv with other models (especially 3B-7B ones), and GRPO loss often does drop "randomly" if the training configuration isn't stable. 

I don't think "there's no reason for double check to be so much worse than gsm8k simple" is true, Qwen2.5-Math-1.5B-Instruct is instruction-tuned only for single-turn CoT Q&A, and DoubleCheck with few-shot prompting is a good bit out-of-distribution for it. RL is messy and runs can go wrong for lots of non-obvious reasons. As a general rule, you should expect that GRPO runs will often "fail" for reasons that aren't easy to pin down.

If there is a particular part of the code that you suspect may be causing issues, I'm happy to take a closer look.
